polaris/hub/client.py

```diff
@@ -623,6 +632,85 @@ def _upload_dataset(
 
             return response
 
+    def _upload_v2_dataset(
+        self,
+        dataset: DatasetV2,
+        access: AccessType = "private",
+        timeout: TimeoutTypes = (10, 200),
+        owner: HubOwner | str | None = None,
+        if_exists: ZarrConflictResolution = "replace",
+    ):
+        """
+        Upload a V2 dataset to the Polaris Hub.
+        """
+
+        with ProgressIndicator(
+            start_msg="Uploading artifact...",
+            success_msg="Uploaded artifact.",
+            error_msg="Failed to upload dataset.",
+        ) as progress_indicator:
+            # Check if a dataset license was specified prior to upload
+            if not dataset.license:
+                raise InvalidDatasetError(
+                    f"\nPlease specify a supported license for this dataset prior to uploading to the Polaris Hub.\nOnly some licenses are supported - {get_args(SupportedLicenseType)}."
+                )
+
+            # Normalize timeout
+            if timeout is None:
+                timeout = self.settings.default_timeout
+
+            # Get the serialized data-modelx (excluding the user-specific cache-dir)
+            dataset.owner = HubOwner.normalize(owner or dataset.owner)
+            dataset_json = dataset.model_dump(exclude_none=True, by_alias=True)
+
+            # Upload dataset meta-data
+            url = f"/v2/dataset/{dataset.owner}/{dataset.name}"
+            response = self._base_request_to_hub(
+                url=url,
+                method="PUT",
+                json={
+                    "zarrManifestFileContent": {
+                        "md5Sum": dataset.md5sum,
+                    },
+                    "access": access,
+                    **dataset_json,
+                },
+                timeout=timeout,
+            )
+
+            with StorageSession(self, "write", dataset.urn) as storage:
+                # Step 2: Upload the manifest file
+                logger.info("Copying the dataset manifest file to the Hub.")
+                with open(dataset.zarr_manifest_path, "rb") as manifest_file:
```

Comment:
Not worth changing now, but this may be a bottleneck if the dataset, and thus the manifest, is sufficiently large.

We can leverage some [useful methods](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetFile.html#pyarrow.parquet.ParquetFile.iter_batches) in PyArrow to bypass reading the entire manifest at once when/if this becomes an issue.

Reply:
Good point that. I think I have an idea for that, but it'll wait until the next PR.